File: examples/scala/classic-driver.scala
/* Driver to run classic driver code */

object Classic extends App {
  var simple_tree : M_SIMPLE = null;
  var p: Any = _;
  if (args.length == 0) {
    simple_tree = new M_SIMPLE("Simple");
    val t_simple = simple_tree.t_Result;
    val ds =
	t_simple.v_xcons_decls(t_simple.v_no_decls(),
			       t_simple.v_decl("x",t_simple.v_integer_type()));
    val s =
	t_simple.v_assign_stmt(t_simple.v_variable("x"),
			       t_simple.v_variable("y"));
    val ss = t_simple.v_xcons_stmts(t_simple.v_no_stmts(),s);
    p = t_simple.v_program(t_simple.v_block(ds,ss));
  } else {
    val ss = new SimpleScanner(new java.io.FileReader(args(0)));
    val sp = new SimpleParser();
    sp.reset(ss, args(0));
    if (!sp.yyparse()) {
      println("Errors found.\n");
      System.exit(1);
    }
    simple_tree = sp.getTree();
    p = simple_tree.t_Program;
  }
  
  val m_simple = simple_tree;
  val m_binding = new M_NAME_RESOLUTION[m_simple.T_Result]("Binding",m_simple.t_Result);
  val t_binding = m_binding.t_Result;

  Debug.activate();

  m_simple.finish();
  m_binding.finish();
  println("Messages:");

  var program: t_binding.T_Program = p.asInstanceOf[t_binding.T_Program];
  for (m <- t_binding.v_prog_msgs(program)) {
    println(m);
  }
}

File: examples/cpp/classic-driver.cpp
#include "aps-impl.h"
#include "basic.h"
#include "simple.h"
#include "classic-binding.h"

#include <iostream>
#include <stdexcept>

using namespace std;

typedef C_NAME_RESOLUTION<C_SIMPLE> BINDING;

void print_bag(C_BAG<C_String>* type, C_BAG<C_String>::T_Result a_bag)
{
  if (a_bag->cons == type->c_append) {
    C_BAG<C_String>::V_append *n = (C_BAG<C_String>::V_append*)a_bag;
    print_bag(type,n->v_l1);
    print_bag(type,n->v_l2);
  } else if (a_bag->cons == type->c_single) {
    C_BAG<C_String>::V_single *n = (C_BAG<C_String>::V_single*)a_bag;
    cout << n->v_x;
  } else if (a_bag->cons == type->c_none) {
  } else {
    cout << "Badly typed?\n";
  }
}

int main()
{
  C_SIMPLE* simple = new C_SIMPLE();
  C_SIMPLE::T_Decls ds =
    simple->v_xcons_decls(simple->v_no_decls(),
			  simple->v_decl("x",simple->v_integer()));
  C_SIMPLE::T_Stmt s =
    simple->v_assign_stmt(simple->v_variable("x"),simple->v_variable("y"));
  C_SIMPLE::T_Stmts ss = simple->v_xcons_stmts(simple->v_no_stmts(),s);
					    
  C_SIMPLE::T_Program p = simple->v_program(simple->v_block(ds,ss));

  Debug::out(cout);

  try {
    BINDING* binding =
      new BINDING(simple);
    
    binding->finish();
    cout << "Messages:" << endl;
    print_bag(binding->t_Messages,binding->v_prog_msgs(p));
  } catch (exception& e) {
    cerr << "Exception: " << e.what() << endl;
  }
}
